src/hotspot/share/gc/g1/g1YoungGCPostEvacuateTasks.hpp

```diff
@@ -0,0 +1,199 @@
+
+  double worker_cost() const override;
+  void do_work(uint worker_id) override;
+};
+
+class G1PostEvacuateCollectionSetCleanupTask1::RemoveSelfForwardPtrsTask : public G1AbstractSubTask {
+  G1ParRemoveSelfForwardPtrsTask _cl;
+
+public:
+  RemoveSelfForwardPtrsTask(G1RedirtyCardsQueueSet* rdcqs);
+  ~RemoveSelfForwardPtrsTask();
+
+  static bool should_execute();
+
+  double worker_cost() const override;
+  void do_work(uint worker_id) override;
+};
+
+// Second set of post evacuate collection set tasks containing (s means serial):
+// - Eagerly Reclaim Humongous Objects (s)
+// - Purge Code Roots (s)
+// - Reset Hot Card Cache (s)
+// - Update Derived Pointers (s)
+// - Redirty Logged Cards
+// - Restore Preserved Marks (on evacuation failure)
+// - Free Collection Set
+class G1PostEvacuateCollectionSetCleanupTask2 : public G1BatchedGangTask {
+  class EagerlyReclaimHumongousObjectsTask;
+  class PurgeCodeRootsTask;
+  class ResetHotCardCacheTask;
+#if COMPILER2_OR_JVMCI
+  class UpdateDerivedPointersTask;
+#endif
+
+  class RedirtyLoggedCardsTask;
+  class RestorePreservedMarksTask;
+  class FreeCollectionSetTask;
+
+public:
+  G1PostEvacuateCollectionSetCleanupTask2(PreservedMarksSet* preserved_marks_set,
+                                          G1RedirtyCardsQueueSet* rdcqs,
+                                          G1EvacuationInfo* evacuation_info,
+                                          const size_t* surviving_young_words);
+};
+
+class G1PostEvacuateCollectionSetCleanupTask2::ResetHotCardCacheTask : public G1AbstractSubTask {
+public:
+  ResetHotCardCacheTask() : G1AbstractSubTask(G1GCPhaseTimes::ResetHotCardCache) { }
+
+  double worker_cost() const override { return 0.5; }
+  void do_work(uint worker_id) override;
+};
+
+class G1PostEvacuateCollectionSetCleanupTask2::PurgeCodeRootsTask : public G1AbstractSubTask {
+public:
+  PurgeCodeRootsTask() : G1AbstractSubTask(G1GCPhaseTimes::PurgeCodeRoots) { }
+
+  double worker_cost() const override { return 1.0; }
+  void do_work(uint worker_id) override;
+};
+
+#if COMPILER2_OR_JVMCI
+class G1PostEvacuateCollectionSetCleanupTask2::UpdateDerivedPointersTask : public G1AbstractSubTask {
+public:
+  UpdateDerivedPointersTask() : G1AbstractSubTask(G1GCPhaseTimes::UpdateDerivedPointers) { }
+
+  double worker_cost() const override { return 1.0; }
+  void do_work(uint worker_id) override;
+};
+#endif
+
+class G1PostEvacuateCollectionSetCleanupTask2::EagerlyReclaimHumongousObjectsTask : public G1AbstractSubTask {
+  uint _humongous_regions_reclaimed;
+  size_t _bytes_freed;
+
+public:
+  EagerlyReclaimHumongousObjectsTask();
+  virtual ~EagerlyReclaimHumongousObjectsTask();
+
+  static bool should_execute();
+
+  double worker_cost() const override { return 1.0; }
+  void do_work(uint worker_id) override;
+};
+
+class G1PostEvacuateCollectionSetCleanupTask2::RestorePreservedMarksTask : public G1AbstractSubTask {
+  PreservedMarksSet* _preserved_marks;
+  AbstractGangTask* _cl;
```

Comment:
Again, I think `_task` makes more sense. Especially since it is initialized with `create_task()`.